core/cache/model.go

```diff
@@ -146,6 +149,26 @@ func (m *Model) Machine(machineId string) (*Machine, error) {
 	return machine, nil
 }
 
+// Branch returns the branch with the input name.
+// If the branch is not found, a NotFoundError is returned.
+// All API-level logic identifies active branches by their name whereas they
+// are managed in the cache by ID - we iterate over the map to locate them.
+// We do not expect many active branches to exist at once,
+// so the performance should be acceptable.
+// TODO (manadart 2019-05-30): Note that this returns a copy of the branch.
+// All of the other entity retrieval should be changed to work in the same
+// fashion, and the lock guarding of member access removed where possible.
+func (m *Model) Branch(name string) (Branch, error) {
+	defer m.doLocked()()
+
+	for _, b := range m.branches {
+		if b.details.Name == name {
+			return *b, nil
```

Comment:
While this does return a copy of the Branch, the BranchChange struct contains 2 maps that will be shallow-copied. Can this become a problem?

I lack extra context on the inner workings of Resident but won't  calling `m.Branch("foo")` multiple times generate  branches copies that share the same `*Resident` reference? Is that safe to share?

Reply:
The shallow copy could be a problem. I will fix that.

The resident reference is actually required for creating watchers directly from the branch copy. It should be safe, because the only thing exported from `Resident` is the read-only `CacheId` method.